File: Codigo proyecto 2/src/motor.cpp
#include <Arduino.h>
#include "motor.h"

Motor::Motor(int _enA, int _in1, int _in2, int _enB, int _in3, int _in4) {
    enA = _enA; in1 = _in1; in2 = _in2;
    enB = _enB; in3 = _in3; in4 = _in4;
    velocidad = 50;  // Velocidad inicial 110
    estado = 0;       // Estado inicial: detenido
}

void Motor::configurar() {
    pinMode(enA, OUTPUT);
    pinMode(in1, OUTPUT);
    pinMode(in2, OUTPUT);
    pinMode(enB, OUTPUT);
    pinMode(in3, OUTPUT);
    pinMode(in4, OUTPUT);
    Serial.println("[Motor] Motores configurados correctamente.");
}

void Motor::avanzar(int vel) {
    if (vel < 1 || vel > 255) {
        Serial.printf("[Motor] Error: Velocidad inválida (Avanzar) %d\n", vel);
        return;
    }
    Serial.printf("[Motor] Avanzar: ENA=%d, ENB=%d\n", vel, vel);
    
    //velocidad = constrain(vel, 110, 255);
    estado = 1;
    
    analogWrite(enA, velocidad);
    analogWrite(enB, velocidad);
    digitalWrite(in1, LOW);
    digitalWrite(in2, HIGH);
    digitalWrite(in3, HIGH);
    digitalWrite(in4, LOW);
}

void Motor::retroceder(int vel) {
    if (vel < 110 || vel > 255) {
        Serial.printf("[Motor] Error: Velocidad inválida (Retroceder) %d\n", vel);
        return;
    }
    Serial.printf("[Motor] Retroceder: ENA=%d, ENB=%d\n", vel, vel);
    
    //velocidad = constrain(vel, 110, 255);
    estado = 2;
    analogWrite(enA, velocidad);
    analogWrite(enB, velocidad);
    digitalWrite(in1, HIGH);
    digitalWrite(in2, LOW);
    digitalWrite(in3, LOW);
    digitalWrite(in4, HIGH);
}

void Motor::girarIzquierda(int vel) {
    if (vel < 110 || vel > 255) {
        Serial.printf("[Motor] Error: Velocidad inválida (Girar Izquierda) %d\n", vel);
        return;
    }
    int velDer = vel;
    int velIzq = vel;  // Reducción del 40%
    Serial.printf("[Motor] Girar Izquierda: ENA=%d, ENB=%d\n", velDer, velIzq);
    
    //velocidad = constrain(vel, 110, 255);
    estado = 3;
    analogWrite(enA, velDer);
    analogWrite(enB, velIzq);
    digitalWrite(in1, HIGH);
    digitalWrite(in2, LOW);
    digitalWrite(in3, HIGH);
    digitalWrite(in4, LOW);
}

void Motor::girarDerecha(int vel) {
    if (vel < 1 || vel > 255) {
        Serial.printf("[Motor] Error: Velocidad inválida (Girar Derecha) %d\n", vel);
        return;
    }
    int velIzq = vel;
    int velDer = vel;  // Reducción del 40%
    Serial.printf("[Motor] Girar Derecha: ENA=%d, ENB=%d\n", velDer, velIzq);
    
    //velocidad = constrain(vel, 110, 255);
    estado = 4;
    analogWrite(enA, velDer);
    analogWrite(enB, velIzq);
    digitalWrite(in1, LOW);
    digitalWrite(in2, HIGH);
    digitalWrite(in3, LOW);
    digitalWrite(in4, HIGH);
}

void Motor::detener() {
    Serial.println("[Motor] Deteniendo motores...");
    estado = 0;
    analogWrite(enA, 0);
    analogWrite(enB, 0);
    digitalWrite(in1, LOW);
    digitalWrite(in2, LOW);
    digitalWrite(in3, LOW);
    digitalWrite(in4, LOW);
}

void Motor::setVelocidad(int vel) {
    if (vel < 110 || vel > 255) {
        Serial.printf("[Motor] Error: Velocidad inválida (Set Velocidad) %d\n", vel);
        return;
    }
    //velocidad = constrain(vel, 110, 255);
    Serial.printf("[Motor] Nueva velocidad: %d\n", velocidad);
    actualizarVelocidad();
}

int Motor::getVelocidad() {
    return velocidad;
}

void Motor::actualizarVelocidad() {
    if (estado != 0) {  // Si los motores están en movimiento
        Serial.printf("[Motor] Actualizando velocidad: Estado=%d, Velocidad=%d\n", estado, velocidad);
        switch (estado) {
            case 1: avanzar(velocidad); break;
            case 2: retroceder(velocidad); break;
            case 3: girarIzquierda(velocidad); break;
            case 4: girarDerecha(velocidad); break;
        }
    }
}

File: Codigo proyecto 2/motor.py
# motor.py (Modificado para usar PWM de 16 bits directamente)
from machine import Pin, PWM

class Motor:
    def __init__(self, en_a, in1, in2, en_b, in3, in4):
        # Configuración de pines PWM y de dirección
        self.enA = PWM(Pin(en_a))
        self.enA.freq(1000)  # Frecuencia PWM
        self.in1 = Pin(in1, Pin.OUT)
        self.in2 = Pin(in2, Pin.OUT)

        self.enB = PWM(Pin(en_b))
        self.enB.freq(1000)  # Frecuencia PWM
        self.in3 = Pin(in3, Pin.OUT)
        self.in4 = Pin(in4, Pin.OUT)

        # Inicializar motores detenidos
        self.detener()

    def set_pwm_duties(self, duty_A, duty_B):
        """
        Establece directamente los ciclos de trabajo PWM (0-65535) para los motores A y B.
        """
        # Asegurarse que los valores están en el rango correcto (0-65535 para duty_u16)
        duty_A_u16 = max(0, min(65535, int(duty_A)))
        duty_B_u16 = max(0, min(65535, int(duty_B)))
        # Aplicar los ciclos de trabajo
        self.enA.duty_u16(duty_A_u16)
        self.enB.duty_u16(duty_B_u16)

    def avanzar(self, duty_cycle):
        """
        Mueve ambos motores hacia adelante con el ciclo de trabajo especificado (0-65535).
        """
        # Establecer dirección para avanzar
        self.in1.value(0)
        self.in2.value(1)
        self.in3.value(1) # Asumiendo cableado estándar para avanzar
        self.in4.value(0)
        # Establecer la velocidad (ciclo de trabajo)
        self.set_pwm_duties(duty_cycle, duty_cycle)

    def retroceder(self, duty_cycle):
        """
        Mueve ambos motores hacia atrás con el ciclo de trabajo especificado (0-65535).
        """
        # Establecer dirección para retroceder
        self.in1.value(1)
        self.in2.value(0)
        self.in3.value(0) # Asumiendo cableado estándar para retroceder
        self.in4.value(1)
        # Establecer la velocidad (ciclo de trabajo)
        self.set_pwm_duties(duty_cycle, duty_cycle)

    def girar_izquierda(self, duty_cycle):
        """
        Gira el robot hacia la izquierda (motor A atrás, motor B adelante)
        con el ciclo de trabajo especificado (0-65535).
        """
        # Establecer dirección para girar izquierda
        self.in1.value(1); self.in2.value(0) # Motor A atrás
        self.in3.value(1); self.in4.value(0) # Motor B adelante
        # Establecer la velocidad (ciclo de trabajo)
        self.set_pwm_duties(duty_cycle, duty_cycle)

    def girar_derecha(self, duty_cycle):
        """
        Gira el robot hacia la derecha (motor A adelante, motor B atrás)
        con el ciclo de trabajo especificado (0-65535).
        """
        # Establecer dirección para girar derecha
        self.in1.value(0); self.in2.value(1) # Motor A adelante
        self.in3.value(0); self.in4.value(1) # Motor B atrás
        # Establecer la velocidad (ciclo de trabajo)
        self.set_pwm_duties(duty_cycle, duty_cycle)

    def detener(self):
        """
        Detiene ambos motores (PWM a 0 y pines de dirección a 0).
        """
        self.set_pwm_duties(0, 0) # Poner ciclo de trabajo a 0
        # Poner pines de dirección a bajo (puede ayudar como freno o simplemente apagar)
        self.in1.value(0); self.in2.value(0)
        self.in3.value(0); self.in4.value(0)

